packages/realm/src/app-services/SyncSession.ts

```diff
@@ -229,106 +228,90 @@ const mockApp = new Proxy({} as App, {
 const PROGRESS_LISTENERS = new Listeners<
   ProgressNotificationCallback,
   ListenerToken,
-  [binding.SyncSession, ProgressDirection, ProgressMode]
+  [binding.WeakSyncSession, binding.SyncSession, ProgressDirection, ProgressMode]
 >({
   throwOnReAdd: true,
-  add(callback, internal, direction, mode) {
+  add(callback, weakInternal, internal, direction, mode) {
     const token = internal.registerProgressNotifier(
       (transferred, transferable) => callback(Number(transferred), Number(transferable)),
       toBindingDirection(direction),
       mode === ProgressMode.ReportIndefinitely,
     );
-    return { internal, token };
+    return { weakInternal, token };
   },
-  remove({ internal, token }) {
-    return internal.unregisterProgressNotifier(token);
+  remove({ weakInternal, token }) {
+    weakInternal.withDeref((internal) => internal?.unregisterProgressNotifier(token));
```

Comment:
Note: unlike other cases, this will *NOT* throw if the Session has been destroyed. `removing()` a listener from a destroyed session will silently do nothing. This seams reasonable to me since the listener was removed when the session was destroyed, and because `remove()` seems likely to be called as part of teardown, which may have caused the session to be destroyed already.

Reply:
I'm a bit curious on why `internal` is able to be undefined / null, i.e. why you need the `?.` here.

Reply:
Do you think it is worth a comment like `// Note: not using the withInternal() helper, so internal can be null`?

Reply:
Nah - I just had to take a step back and read it properly 👍